### PR TITLE
Fixes related to File API

### LINKED ISSUE #10851: Add file handling functionality to Blob object
**NOTE**: This is an issue assigned to myself

Spec: https://w3c.github.io/FileAPI/#blob

As noted before, current Blob object can only refer to in-memory data ([`DataSlice`](https://doc.servo.org/script/dom/blob/struct.DataSlice.html)), and the [`FileReader`](https://doc.servo.org/script/dom/filereader/struct.FileReader.html) will also only operate in this way. In order to support reading from file system, another internal data structure might need to be added to the `Blob`. It might be a file handler created at the time of object construction.

About snapshot state:

> If a File object is a reference to a byte sequence originating from a file on disk, then its snapshot state should be set to the state of the file on disk at the time the File object is created.
> 
> Note: This is a non-trivial requirement to implement for user agents, and is thus not a must but a should [RFC2119]. User agents should endeavor to have a File object’s snapshot state set to the state of the underlying storage on disk at the time the reference is taken. If the file is modified on disk following the time a reference has been taken, the File's snapshot state will differ from the state of the underlying storage. User agents may use modification time stamps and other mechanisms to maintain snapshot state, but this is left as an implementation detail.

Working at [branching here](https://github.com/izgzhen/servo/tree/blob)

### PR DESCRIPTION
**Pull request by izgzhen**
Fixing problems I met when trying to resolve https://github.com/servo/servo/issues/10851, but most of changes here are simple fixes.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10873)

### CODE REVIEW

**Comment by izgzhen** [comment]
@Ms2ger Looks like the changed `Cargo.lock` triggered CI failure. How to deal with this?

**Comment by Ms2ger** [comment]
It's not the one you changed, it's the ones you didn't change. `./mach cargo-update -p script` should help.

**Comment by izgzhen** [comment]
I simplified the content of this contrib, since my discussion with @Manishearth shows that there are some design decisions to be settled down later.

Welcome review.

File: components/script/dom/file.rs [hunk 1]
```diff
@@ -18,14 +20,15 @@ pub struct File {
 }
 
 impl File {
-    fn new_inherited(_file_bits: &Blob, name: DOMString) -> File {
+    fn new_inherited(file_bits: &Blob, name: DOMString) -> File {
+       let mime_type = guess_mime_type(Path::new(&*name));
```
Comment (Manishearth) [inline components/script/dom/file.rs]: Why are we guessing the type here? The spec doesn't mention this, the type is the empty string unless explicitly provided by the user through the FilePropertyBag/BlobPropertyBag.

**Comment by izgzhen** [comment]
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/file.rs, line 24 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10873#-KGLqxxYQ1OUfi_wsv_o-r6-24:-KGjjUki4MQRpTzH2sjV:1010503494) ([raw file](https://github.com/servo/servo/blob/f1a5451807b5c41ad2f6d28f7b1e18d5973d31c5/components/script/dom/file.rs#L24)):*
Good point! There was a "//TODO: get type from the underlying filesystem instead of "".to_string()" left in the initial version and I kind of get mislead by this.

This snippet should be used in the internal initial construction when the files are selected by user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10873)*

**Comment by Manishearth** [comment]
@bors-servo r+

---

Reviewed 2 of 12 files at r6, 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/file.rs, line 24 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10873#-KGLqxxYQ1OUfi_wsv_o-r6-24:-KGjs4RgoSSFmfOCceDq:-705065685) ([raw file](https://github.com/servo/servo/blob/f1a5451807b5c41ad2f6d28f7b1e18d5973d31c5/components/script/dom/file.rs#L24)):*
Yeah, that was wrong :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10873#-:-KGjsZRjl0XthWbgf--d:1236647940)*

**Comment by izgzhen** [comment]
Ping issue #11131

**Pull request merged**